### PR TITLE
CBL-445: After doc rejected in push, doesn't pull newer revisions

### PR DESCRIPTION
**Pull request by jianminzhao**
We will remember if a doc is rejected in push, and provide a new API, c4doc_isRevRejected() to query it. This function is designed to aid DocumentEndedCallback in resolving conflicts with the current revision that was rejected.

### CODE REVIEW

File: LiteCore/Database/CollectionImpl.hh [hunk 1]
```diff
@@ -210,7 +210,11 @@ namespace litecore {
                 } while ( doc->selectNextRevision() );
                 if ( !revID ) return false;
             }
-            doc->setRemoteAncestorRevID(remoteID, revID);
+            if ( remoteID == RevTree::kNoRemoteID ) {
+                doc->revIsRejected(revID);
```
Comment (pasin) [inline LiteCore/Database/CollectionImpl.hh:214]: Why need to call this function here?

File: LiteCore/RevTrees/RawRevTree.cc [hunk 2]
```diff
@@ -64,23 +65,38 @@ namespace litecore {
         while ( entry < (const void*)raw_tree.end() ) {
             RevTree::RemoteID remoteID = endian::dec16(entry->remoteDBID_BE);
             auto              revIndex = endian::dec16(entry->revIndex_BE);
+            if ( remoteID == 0 && revIndex == 0 ) {
+                ++entry;
+                break; // The zero mark
```
Comment (pasin) [inline LiteCore/RevTrees/RawRevTree.cc:70]: I don't know about the structure of the rev-tree. What is the zero mark? Is this new?

File: LiteCore/RevTrees/RawRevTree.cc [hunk 3]
```diff
@@ -64,23 +65,38 @@ namespace litecore {
...
+                error::_throw(error::CorruptRevisionData, "RawRevision dcodeTree revIndex error at remoteMap");
             remoteMap[remoteID] = &revs[revIndex];
             ++entry;
         }
 
+        while ( entry < (const void*)raw_tree.end() ) {
+            RevTree::RemoteID remoteID = endian::dec16(entry->remoteDBID_BE);
+            auto              revIndex = endian::dec16(entry->revIndex_BE);
+            if ( remoteID != 0 || revIndex >= count )
+                error::_throw(error::CorruptRevisionData, "RawRevision dcodeTree revIndex error at rejectedRevs");
```
Comment (pasin) [inline LiteCore/RevTrees/RawRevTree.cc:82]: dcodeTree (typo)

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 4]
```diff
@@ -2244,3 +2245,202 @@ TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Give SG a rev history with a gap",
         REQUIRE(revID.hasPrefix(to_string(numRevs1 + numInitialRevs) + "-"));
     }
 }
+
+// To run this test, the sync function for Tulips collection must be,
+// "tulips":{"sync":"function(doc,olddoc){if(doc.cbl_445==\"bad\")throw({\"forbidden\":\"read_only\"});channel(doc.channels)}"}
```
Comment (pasin) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:2250]: Maybe create a new collection with this sync function? So SG config doesn't need to be updated before running this test.

Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:2250]: This sync function wouldn't hurt other tests. I used persistent config. I don't know if I can add a new collection after collections are registered with PUT a new db.

File: LiteCore/Database/CollectionImpl.hh (hunk 1 above)
Comment (jianminzhao) [reply LiteCore/Database/CollectionImpl.hh:214]: c.f. Pusher+Revs.cc, where we received 403 from SG. There, we don't have Rev object itself, we only work with ReplicatedRev data structure where Rev is represented as RevID. This is the place where we do book-keeping as the rev is finally synced. In this case, it is kind-of "synced" in the sense that it is formally rejected by SG.
That is, as a rev comes to here, either, it is normally synced, and we mark the remote ancestor. Or, it is abnormally synced, or formally rejected, we remember this fact.

File: Replicator/tests/ReplicatorCollectionSGTest.cc (hunk 4 above)
Comment (pasin) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:2250]: You can add another collection in the PUT request.

Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:2250]: {"error":"Precondition Failed","reason":"Duplicate database name \"scratch\""}

File: LiteCore/Database/CollectionImpl.hh (hunk 1 above)
Comment (pasin) [reply LiteCore/Database/CollectionImpl.hh:214]: It's just a weird to call the getter method to do something here.

Comment (pasin) [reply LiteCore/Database/CollectionImpl.hh:214]: Wait, it's not the getter. Sorry, I understood this wrong.

File: LiteCore/RevTrees/RawRevTree.cc (hunk 2 above)
Comment (borrrden) [reply LiteCore/RevTrees/RawRevTree.cc:70]: This definitely needs to be documented in comments.  Don't just add something artificial without an explanation.

**Review by borrrden — CHANGES_REQUESTED** [review]

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 5]
```diff
@@ -2246,13 +2246,14 @@ TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Give SG a rev history with a gap",
     }
 }
 
-// To run this test, the sync function for Tulips collection must be,
-// "tulips":{"sync":"function(doc,olddoc){if(doc.cbl_445==\"bad\")throw({\"forbidden\":\"read_only\"});channel(doc.channels)}"}
-TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Use isRevRejected to Resolve Conflict", "[.SyncServerCollection]") {
+// This test uses collection, "cbl443", which has the following sync function:
+// "cbl445":{"sync":"function(doc,olddoc){if(doc.cbl_445==\"bad\")throw({\"forbidden\":\"read_only\"});channel(doc.channels)}"}
+TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Use isRevRejected to Resolve Conflict", "[.xSyncServerCollection]") {
```
Comment (pasin) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:2251]: Can you update here as well?

https://github.com/couchbase/couchbase-lite-core/blob/2f5430f6a7b6644b7b21f557e7f1f33282270aff/Replicator/tests/ReplicatorCollectionSGTest.cc#L20-L63

Maybe use another flower name such as "marigold" for consistency.

Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:2251]: Why another flower? What would be another flower different from Tulips which I used originally

**Review by pasin — APPROVED** [review]

**Review by borrrden — APPROVED** [review]

**Pull request merged**